src/java_bytecode/java_object_factory.cpp

```diff
@@ -494,6 +495,141 @@ class recursion_set_entryt
+  const side_effect_expr_nondett nondet_length(length_expr.type());
+  code.add(code_declt(length_expr));
+  code.add(code_assignt(length_expr, nondet_length));
+
+  // assume (length_expr >= 0);
+  code.add(code_assumet(binary_relation_exprt(
+    length_expr, ID_ge, from_integer(0, java_int_type()))));
+
+  // assume (length_expr <= max_input_length)
+  if(max_nondet_string_length<=max_value(length_expr.type()))
+  {
+    exprt max_length=from_integer(max_nondet_string_length, length_expr.type());
+    code.add(code_assumet(
+      binary_relation_exprt(length_expr, ID_le, max_length)));
+  }
+
+  // data_expr = nondet(char[INFINITY]) // we use infinity for variable size
+  exprt data_expr=make_nondet_infinite_char_array(symbol_table, loc, code);
+
+  // @clsid = String and @lock = false:
+  const symbol_typet jlo_symbol("java::java.lang.Object");
+  const struct_typet jlo_type=to_struct_type(ns.follow(jlo_symbol));
+  struct_exprt jlo_init(jlo_symbol);
+  const constant_exprt class_id_expr(class_id, jlo_type.components()[0].type());
+  jlo_init.copy_to_operands(class_id_expr);
+  jlo_init.copy_to_operands(
+    from_integer(false, jlo_type.components()[1].type()));
+
+  struct_exprt struct_expr(obj.type());
+  struct_expr.copy_to_operands(jlo_init);
+  struct_expr.copy_to_operands(length_expr);
+
+  const address_of_exprt first_index(index_exprt(
+       data_expr, from_integer(0, java_int_type())));
+  struct_expr.copy_to_operands(first_index);
+
+  add_pointer_to_array_association(
+    first_index, data_expr, symbol_table, loc, code);
+
+  add_array_to_length_association(
+    data_expr, length_expr, symbol_table, loc, code);
+
+  // tmp_object = struct_expr;
+  code.add(code_assignt(obj, struct_expr));
+  return code;
+}
+
+/// Add code for the initialization of a string and affectation of its
+/// address to the pointer `expr`
+/// \param expr: pointer to be affected
+/// \param symbol_table: the symbol table
+/// \param loc: location in the source
+/// \param [out] code: code block in which initialization code is added
+static void add_nondet_string_pointer_initialization(
+  const exprt &expr,
+  const std::size_t &max_nondet_string_length,
+  symbol_tablet &symbol_table,
+  const source_locationt &loc,
+  code_blockt &code)
+{
+  const namespacet ns(symbol_table);
+  const dereference_exprt obj(expr, expr.type().subtype());
+  const struct_typet struct_type=
+      (obj.type().id()!=ID_struct)?
+        to_struct_type(ns.follow(to_symbol_type(obj.type()))):
+        to_struct_type(obj.type());
+  const symbolt tmp_object_sym=new_tmp_symbol(symbol_table, loc, struct_type);
+  const symbol_exprt tmp_object=tmp_object_sym.symbol_expr();
+  // struct java.lang.String tmp_object;
+  code.add(code_declt(tmp_object));
+  code.add(initialize_nondet_string_struct(
+    tmp_object, max_nondet_string_length, loc, symbol_table));
+  // expr = &tmp_object;
+  code.add(code_assignt(expr, address_of_exprt(tmp_object), loc));
```

Comment:
Aren't temporaries like this allocated function-local? If so won't it go out of scope too early? Consider stealing the nondet-object-init code in the object factory, which uses dynamic alloc for this.

Reply:
@smowton 
Since this function `add_nondet_string_pointer_initialization` is only used in `__CPROVER_start` I was thinking that it wouldn't be a problem (but I may be wrong). Here is an example of what `__CPROVER_start` looks like with a string initialized with this function (`tmp_object_factory$6` is the tmp object we take the address of here):
```c
__CPROVER__start /* __CPROVER__start */
        // 258 no location
        __CPROVER_initialize();
        // 259 no location
        uncaught_exception = null;
        // 260 file test.java line 5 function java::test.check:(Ljava/lang/String;)[C
        struct java.lang.String *arg0a;
        // 261 no location
        struct java.lang.String { struct java.lang.Object @java.lang.Object; int length; char *data; } tmp_object_factory$6;
        // 262 no location
        int tmp_object_factory$7;
        // 263 no location
        tmp_object_factory$7 = NONDET(int);
        // 264 no location
        ASSUME tmp_object_factory$7 >= 0
        // 265 no location
        ASSUME tmp_object_factory$7 <= 2147483647
        // 266 no location
        char nondet_infinite_array$8[INFINITY()];
        // 267 no location
        nondet_infinite_array$8 = NONDET(char [INFINITY()]);
        // 268 no location
        int return_array$9;
        // 269 no location
        return_array$9 = cprover_associate_array_to_pointer_func(nondet_infinite_array$8, nondet_infinite_array$8);
        // 270 no location
        int return_array$10;
        // 271 no location
        return_array$10 = cprover_associate_length_to_array_func(nondet_infinite_array$8, tmp_object_factory$7);
        // 272 no location
        tmp_object_factory$6 = { .@java.lang.Object={ .@class_identifier="java.lang.String", .@lock=false },
    .length=tmp_object_factory$7, .data=nondet_infinite_array$8 };
        // 273 no location
        dead return_array$10;
        // 274 no location
        dead return_array$9;
        // 275 no location
        dead nondet_infinite_array$8;
        // 276 no location
        dead tmp_object_factory$7;
        // 277 file test.java line 5 function java::test.check:(Ljava/lang/String;)[C
        arg0a = &tmp_object_factory$6;
        // 278 file test.java line 5 function java::test.check:(Ljava/lang/String;)[C
        INPUT("arg0a", arg0a);
        // 279 file test.java line 5 function java::test.check:(Ljava/lang/String;)[C
        test.check:(Ljava/lang/String;)[C(arg0a);
        // 280 file test.java line 5 function java::test.check:(Ljava/lang/String;)[C
        return = test.check:(Ljava/lang/String;)[C#return_value;
        // 281 file test.java line 5 function java::test.check:(Ljava/lang/String;)[C
        dead test.check:(Ljava/lang/String;)[C#return_value;
        // 282 no location
        GOTO 1
        // 283 no location
        // Labels: toplevel_catch
        SKIP
        // 284 no location
        // Labels: after_catch
     1: SKIP
        // 285 file test.java line 5 function java::test.check:(Ljava/lang/String;)[C
        OUTPUT("return", return);
        // 286 file test.java line 5 function java::test.check:(Ljava/lang/String;)[C
        OUTPUT("arg0a", arg0a);
        // 287 file test.java line 5 function java::test.check:(Ljava/lang/String;)[C
        OUTPUT("uncaught_exception'", uncaught_exception);
        // 288 no location
        dead tmp_object_factory$6;
        // 289 no location
        dead arg0a;
        // 290 no location
        END_FUNCTION
```

Reply:
Couldn't this function get called to provide the nondet string return from a stubbed method?

Reply:
Yes you are right, I will try using a dynamic object instead